### PR TITLE
feat(close payment) added fields into transactionDetails

### PR DESCRIPTION
**Pull request by Fernando-Granato**
Depends on https://github.com/pagopa/pagopa-ecommerce-commons/pull/95

#### List of Changes

👉🏻 **CLOSE PAYMENT OK/KO with authorization completation event** 

Changed the logic to forward the close payment to the node. 
The following fields have been added to the `transactionDetails` object in case of close payment **OK** or **KO** after receiving the request for completion of authorization: 

- brand
- paymentMethodName
- clientId
- errorCode
- timestampOperation
- brokerName
- paymentGateway
- pspOnUs

The field `authorizationCode` will be valorized only in case of close payment **OK** otherwise in case of **KO** the `errorCode` will be valorized.

👉🏻 **CLOSE PAYMENT KO with user cancellation request**

in case of closePayment KO for cancellation by the user will be forwarded only new fields available in the `transactionDetail` field in particular: 

- clientId
- timestampOperation

#### Motivation and Context
The change was made to enhance with the correct information the `transactionDetails` field in the closePayment request to the node 🚀

#### How Has This Been Tested?
The tests were carried out with junit test and running a flow with ecommerce-local 📈

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by ciuffagianluca — APPROVED** [review]
Great job 🚀

File: pom.xml [hunk 1]
```diff
@@ -320,7 +320,7 @@
 						</goals>
 						<configuration>
 							<inputSpec>
-								https://raw.githubusercontent.com/pagopa/pagopa-infra/main/src/core/api/nodopagamenti_api/nodoPerPM/v2/_swagger.json.tpl
+								https://raw.githubusercontent.com/pagopa/pagopa-infra/CHK-1674-closepaymentV2-update-biz-event/src/core/api/nodopagamenti_api/nodoPerPM/v2/_swagger.json.tpl
```
Comment (ciuffagianluca) [inline pom.xml:323]: Remember to restore main branch 👀

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt [hunk 2]
```diff
@@ -223,4 +118,135 @@ class NodeService(
       is TransactionWithClosureError -> getPaymentTypeCode(tx.transactionAtPreviousState)
       else -> TIPO_VERSAMENTO_CP
     }
+
+  private fun buildClosePaymentForCancellationRequest(
+    transactionWithCancellation: BaseTransactionWithCancellationRequested,
+    transactionOutcome: ClosePaymentRequestV2Dto.OutcomeEnum,
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:124]: This must be always equal to `ClosePaymentRequestV2Dto.OutcomeEnum.KO`, so either:

* check this invariant inside this method
* remove this parameter and replace it with the constant

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt [hunk 3]
```diff
@@ -51,148 +53,42 @@ class NodeService(
...
+                .map { trxPreviousStatus ->
+                  when (transactionOutcome) {
+                    ClosePaymentRequestV2Dto.OutcomeEnum.KO -> {
+                      trxPreviousStatus.fold(
+                        { transactionWithCancellation ->
+                          buildClosePaymentForCancellationRequest(
+                            transactionWithCancellation, transactionOutcome, transactionId)
+                        },
+                        { transactionWithCompletedAuthorization ->
+                          buildClosePaymentRequest(
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:68]: What about:
```suggestion
                          buildAuthorizationCompletedClosePaymentRequest(
```

**Review by giovanniberti — COMMENTED** [review]
Very clean! 👏
Just put a dependency on https://github.com/pagopa/pagopa-infra/pull/1124 and update the PR title before merging 😃

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt (hunk 3 above)
Comment (Fernando-Granato) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt]: Fixed with 👉🏻 f049dc8753d07392d8ca6bf0564403d464e8ce6d

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt (hunk 2 above)
Comment (Fernando-Granato) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt]: Fixed with 👉🏻 3740909e21260868428e86bf4c0fcae1dd87fb71

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt [hunk 4]
```diff
@@ -223,4 +119,134 @@ class NodeService(
...
+      idPSP = authCompleted.transactionAuthorizationRequestData.pspId
+      paymentMethod = authCompleted.transactionAuthorizationRequestData.paymentTypeCode
+      idBrokerPSP = authCompleted.transactionAuthorizationRequestData.brokerName
+      idChannel = authCompleted.transactionAuthorizationRequestData.pspChannelCode
+      this.transactionId = transactionId.value()
+      totalAmount =
+        EuroUtils.euroCentsToEuro(
+          (authCompleted.transactionAuthorizationRequestData.amount.plus(
+            authCompleted.transactionAuthorizationRequestData.fee)))
+      fee = authCompleted.transactionAuthorizationRequestData.fee.toBigDecimal()
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:175]: Use  `EuroUtils.euroCentsToEuro` utils also for fees
As for now you are converting int fee value to BigDecimal without converting from eurocents to euro. So a fee with value 100 will be converted to a BigDecimal valued with 100,00.

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt [hunk 5]
```diff
@@ -223,4 +119,134 @@ class NodeService(
...
+        else null
+      transactionDetails =
+        TransactionDetailsDto().apply {
+          transaction =
+            TransactionDto().apply {
+              this.transactionId = transactionId.value()
+              transactionStatus = getTransactionDetailsStatus(authCompleted)
+              paymentGateway = authCompleted.transactionAuthorizationRequestData.paymentGateway.name
+              fee = authCompleted.transactionAuthorizationRequestData.fee.toBigDecimal()
+              amount = authCompleted.transactionAuthorizationRequestData.amount.toBigDecimal()
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:214]: same as above

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt [hunk 6]
```diff
@@ -223,4 +119,134 @@ class NodeService(
...
+              grandTotal =
+                (authCompleted.transactionAuthorizationRequestData.amount.plus(
+                    authCompleted.transactionAuthorizationRequestData.fee))
+                  .toBigDecimal()
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt:218]: same as above

Comment (Fernando-Granato) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt]: Fixed with 👉🏻 2b272267f67e0112034f62e459f128c4d2e11b73

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt (hunk 5 above)
Comment (Fernando-Granato) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt]: Fixed with 👉🏻 2b272267f67e0112034f62e459f128c4d2e11b73

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt (hunk 4 above)
Comment (Fernando-Granato) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeService.kt]: Fixed with 👉🏻 2b272267f67e0112034f62e459f128c4d2e11b73

File: src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeServiceTests.kt [hunk 7]
```diff
@@ -63,24 +63,48 @@ class NodeServiceTests {
...
+      assertEquals(
+        Transaction.ClientId.CHECKOUT.name,
+        closePaymentRequestCaptor.value.transactionDetails.info.clientId)
+      assertEquals(TIPO_VERSAMENTO_CP, closePaymentRequestCaptor.value.transactionDetails.info.type)
+      assertEquals(
+        closePaymentRequestCaptor.value.transactionDetails.transaction.amount,
+        closePaymentRequestCaptor.value.transactionDetails.transaction.grandTotal)
+      assertNull(closePaymentRequestCaptor.value.transactionDetails.transaction.fee)
+      assertNotNull(closePaymentRequestCaptor.value.transactionDetails.transaction.amount)
+      assertNotNull(closePaymentRequestCaptor.value.transactionDetails.transaction.grandTotal)
```
Comment (pietro-tota) [inline src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeServiceTests.kt:93]:
```suggestion
assertNull(closePaymentRequestCaptor.value.transactionDetails.transaction.fee)
      val paymentNoticesAmount = EuroUtils.euroCentsToEuro(activatedEvent.data.paymentNotices.stream().mapToInt{it.amount}.sum())
      assertEquals(paymentNoticesAmount,closePaymentRequestCaptor.value.transactionDetails.transaction.amount)
      assertEquals(paymentNoticesAmount,closePaymentRequestCaptor.value.transactionDetails.transaction.grandTotal)
```
Perform check that both amount and grandTotal are equals to payment notices amount sum

File: src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeServiceTests.kt [hunk 8]
```diff
@@ -92,12 +116,34 @@ class NodeServiceTests {
...
+        closePaymentRequestCaptor.value.transactionDetails.transaction.transactionStatus)
+      assertEquals(
+        Transaction.ClientId.CHECKOUT.name,
+        closePaymentRequestCaptor.value.transactionDetails.info.clientId)
+      assertEquals(TIPO_VERSAMENTO_CP, closePaymentRequestCaptor.value.transactionDetails.info.type)
+      assertEquals(
+        closePaymentRequestCaptor.value.transactionDetails.transaction.amount,
+        closePaymentRequestCaptor.value.transactionDetails.transaction.grandTotal)
+      assertNull(closePaymentRequestCaptor.value.transactionDetails.transaction.fee)
+      assertNotNull(closePaymentRequestCaptor.value.transactionDetails.transaction.amount)
```
Comment (pietro-tota) [inline src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeServiceTests.kt:144]: Same as above

File: src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeServiceTests.kt (hunk 7 above)
Comment (Fernando-Granato) [reply src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeServiceTests.kt:95]: Fixed with 👉🏻 2610b8c15e08d4983d5171bf6bda99f7251f1956

File: src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeServiceTests.kt (hunk 8 above)
Comment (Fernando-Granato) [reply src/test/kotlin/it/pagopa/ecommerce/eventdispatcher/services/NodeServiceTests.kt:152]: Fixed with 👉🏻 2610b8c15e08d4983d5171bf6bda99f7251f1956

**Review by ciuffagianluca — APPROVED** [review]
Well done! 🚀

**Pull request merged**